### PR TITLE
refactor: conversation event processing AR-2620 - part 1

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

The code paths for adding / removing users to a group conversation are duplicated or not implemented correctly. The `handleMemberJoin` `handleMemberLeave` functions in the `ConversationEventReceiver` is partially duplicated by the `AddMemberToConversationUseCase`, `RemoveMemberFromConversationUseCase` and the `addMember` function on the `ConversationRepository` optimistically inserts members instead of deriving the inserts on the server response,

### Solutions

Break up `ConversationEventReceiver` into multiple `EventHandlers` so that the processing of member join & leave events can be re-used in other code paths.

Extract `ConversationGroupRepository` from the `ConversationRepository` and remove the `MLSConversationRepository` dependency so that `ConversationRepository` can be injected in the `MemberJoinEventHandler` without introducing dependency cycles.

#### Test Coverage

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by vitorhugods — APPROVED** [review]
Nice work! This is SUCH a quality of live improvement for us.

This made me realise that we have no coverage for EventHandlers! What a pity. But to be honest, it was hard to test with so many dependencies and so on.

This PR also improves that. It's easier to test things with a single responsibility :)

Nice, nice!

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt [hunk 1]
```diff
@@ -3,20 +3,22 @@ package com.wire.kalium.logic.feature.conversation
...
+    private val conversationGroupRepository: ConversationGroupRepository
 ) {
 
     suspend operator fun invoke(otherUserId: UserId): CreateConversationResult {
         // TODO: filter out self user from the list (just in case of client bug that leads to self user to be included part of the list)
         return conversationRepository.getOneToOneConversationWithOtherUser(otherUserId)
             .fold({ conversationFailure ->
                 if (conversationFailure is StorageFailure.DataNotFound) {
-                    conversationRepository.createGroupConversation(usersList = listOf(otherUserId))
+                    conversationGroupRepository.createGroupConversation(usersList = listOf(otherUserId))
```
[the remark this answers is not in the document]
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt:21]: And regular groups/conversations. Naming is hard.

**Comment by typfel** [comment]
> Nice work! This is SUCH a quality of live improvement for us.
> 
> This made me realise that we have no coverage for EventHandlers! What a pity. But to be honest, it was hard to test with so many dependencies and so on.
> 
> This PR also improves that. It's easier to test things with a single responsibility :)
> 
> Nice, nice!

I some of them are handled in ConversationEventReceiverTest. That test class needs to be split up as well. I figured it's better to not refactor the tests at the same time. I guess it should be part 3 in this series.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/GetOrCreateOneToOneConversationUseCase.kt:21]: isn't just 'Conversation' would be enough for the naming? It contains the type. So why we don't get rid of group, one-on-one in the naming?

**Pull request merged**